File: conformance/code/individual/src/main/java/com/digit/individual/model/IndividualDTO.java
package com.digit.individual.model;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import java.util.List;
import java.util.Map;

/**
 * Wire-layer shape per the v3 spec. Mirrors Go internal/models/individual_dto.go.
 * Notes: tenantId sourced from header (absent here); rowVersion exposed as "version";
 * active exposed as "isActive"; additionaldetails exposed as "additionalAttributes";
 * dateOfBirth is a string (YYYY-MM-DD on output; flexible parse on input).
 */
@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonPropertyOrder({"id", "individualId", "givenName", "familyName", "otherNames", "dateOfBirth",
        "gender", "age", "mobileNumber", "mobileNumberVerified", "altContactNumber", "email",
        "emailVerified", "locale", "isActive", "fatherName", "husbandName", "photo", "userId",
        "additionalAttributes", "version", "requestId", "auditDetail", "address", "identifiers", "documents"})
public class IndividualDTO {

    private String id = "";
    private String individualId;
    private String givenName;
    private String familyName;
    private String otherNames;
    private String dateOfBirth;
    private String gender;
    private Integer age;
    private String mobileNumber;
    private boolean mobileNumberVerified;
    private String altContactNumber;
    private String email;
    private boolean emailVerified;
    private String locale;
    @JsonProperty("isActive")
    private boolean isActive;
    private String fatherName;
    private String husbandName;
    private String photo;
    private String userId;
    @JsonProperty("additionalAttributes")
    private Map<String, Object> additionalAttributes;
    private int version;
    private String requestId;
    private AuditDetail auditDetail;
    @JsonProperty("address")
    private List<AddressDTO> addresses;
    private List<IdentifierDTO> identifiers;
    private List<DocumentDTO> documents;

    @JsonInclude(JsonInclude.Include.ALWAYS)
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    public String getIndividualId() { return individualId; }
    public void setIndividualId(String individualId) { this.individualId = individualId; }
    public String getGivenName() { return givenName; }
    public void setGivenName(String givenName) { this.givenName = givenName; }
    public String getFamilyName() { return familyName; }
    public void setFamilyName(String familyName) { this.familyName = familyName; }
    public String getOtherNames() { return otherNames; }
    public void setOtherNames(String otherNames) { this.otherNames = otherNames; }
    public String getDateOfBirth() { return dateOfBirth; }
    public void setDateOfBirth(String dateOfBirth) { this.dateOfBirth = dateOfBirth; }
    public String getGender() { return gender; }
    public void setGender(String gender) { this.gender = gender; }
    public Integer getAge() { return age; }
    public void setAge(Integer age) { this.age = age; }
    public String getMobileNumber() { return mobileNumber; }
    public void setMobileNumber(String mobileNumber) { this.mobileNumber = mobileNumber; }
    @JsonInclude(JsonInclude.Include.ALWAYS)
    public boolean isMobileNumberVerified() { return mobileNumberVerified; }
    public void setMobileNumberVerified(boolean mobileNumberVerified) { this.mobileNumberVerified = mobileNumberVerified; }
    public String getAltContactNumber() { return altContactNumber; }
    public void setAltContactNumber(String altContactNumber) { this.altContactNumber = altContactNumber; }
    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }
    @JsonInclude(JsonInclude.Include.ALWAYS)
    public boolean isEmailVerified() { return emailVerified; }
    public void setEmailVerified(boolean emailVerified) { this.emailVerified = emailVerified; }
    public String getLocale() { return locale; }
    public void setLocale(String locale) { this.locale = locale; }
    @JsonProperty("isActive")
    @JsonInclude(JsonInclude.Include.ALWAYS)
    public boolean isActive() { return isActive; }
    public void setActive(boolean active) { isActive = active; }
    public String getFatherName() { return fatherName; }
    public void setFatherName(String fatherName) { this.fatherName = fatherName; }
    public String getHusbandName() { return husbandName; }
    public void setHusbandName(String husbandName) { this.husbandName = husbandName; }
    public String getPhoto() { return photo; }
    public void setPhoto(String photo) { this.photo = photo; }
    public String getUserId() { return userId; }
    public void setUserId(String userId) { this.userId = userId; }
    @JsonProperty("additionalAttributes")
    public Map<String, Object> getAdditionalAttributes() { return additionalAttributes; }
    public void setAdditionalAttributes(Map<String, Object> additionalAttributes) { this.additionalAttributes = additionalAttributes; }
    @JsonInclude(JsonInclude.Include.NON_DEFAULT)
    public int getVersion() { return version; }
    public void setVersion(int version) { this.version = version; }
    public String getRequestId() { return requestId; }
    public void setRequestId(String requestId) { this.requestId = requestId; }
    public AuditDetail getAuditDetail() { return auditDetail; }
    public void setAuditDetail(AuditDetail auditDetail) { this.auditDetail = auditDetail; }
    @JsonProperty("address")
    public List<AddressDTO> getAddresses() { return addresses; }
    public void setAddresses(List<AddressDTO> addresses) { this.addresses = addresses; }
    public List<IdentifierDTO> getIdentifiers() { return identifiers; }
    public void setIdentifiers(List<IdentifierDTO> identifiers) { this.identifiers = identifiers; }
    public List<DocumentDTO> getDocuments() { return documents; }
    public void setDocuments(List<DocumentDTO> documents) { this.documents = documents; }
}

File: conformance/code/individual-go/internal/models/individual_dto.go
package models

// IndividualDTO is the wire-layer shape per the v3 spec.
// Notes:
//   - tenantId is sourced from the X-Tenant-ID header, not the body — absent here.
//   - hashedMobileNumber is internal — absent here.
//   - rowVersion is exposed as "version" per spec.
//   - active is exposed as "isActive" per spec.
//   - additionalDetails (DB column) is exposed as "additionalAttributes" per spec.
//   - auditDetail is readOnly (set by mapper on response; ignored on request).
type IndividualDTO struct {
	ID                   string     `json:"id"`
	IndividualID         string     `json:"individualId,omitempty"`
	GivenName            string     `json:"givenName,omitempty"`
	FamilyName           string     `json:"familyName,omitempty"`
	OtherNames           string     `json:"otherNames,omitempty"`
	DateOfBirth          *Date      `json:"dateOfBirth,omitempty"`
	Gender               string     `json:"gender,omitempty"`
	Age                  *int       `json:"age,omitempty"`
	MobileNumber         string     `json:"mobileNumber,omitempty"`
	MobileNumberVerified bool       `json:"mobileNumberVerified"`
	AltContactNumber     string     `json:"altContactNumber,omitempty"`
	Email                string     `json:"email,omitempty"`
	EmailVerified        bool       `json:"emailVerified"`
	Locale               string     `json:"locale,omitempty"`
	IsActive             bool       `json:"isActive"`
	FatherName           string     `json:"fatherName,omitempty"`
	HusbandName          string     `json:"husbandName,omitempty"`
	Photo                string     `json:"photo,omitempty"`
	UserID               string     `json:"userId,omitempty"`
	AdditionalAttributes JSONB      `json:"additionalAttributes,omitempty"`
	Version              int        `json:"version,omitempty"`
	RequestID            string     `json:"requestId,omitempty"`

	AuditDetail *AuditDetail `json:"auditDetail,omitempty"`

	Addresses   []AddressDTO    `json:"address,omitempty"`
	Identifiers []IdentifierDTO `json:"identifiers,omitempty"`
	Documents   []DocumentDTO   `json:"documents,omitempty"`
}

// IndividualFromEntity converts a DB entity into a wire DTO.
// Flat audit columns are assembled into the nested AuditDetail.
func IndividualFromEntity(e *Individual) *IndividualDTO {
	if e == nil {
		return nil
	}
	d := &IndividualDTO{
		ID:                   e.ID,
		IndividualID:         e.IndividualID,
		GivenName:            e.GivenName,
		FamilyName:           e.FamilyName,
		OtherNames:           e.OtherNames,
		DateOfBirth:          DateFromTimePtr(e.DateOfBirth),
		Gender:               e.Gender,
		Age:                  e.Age,
		MobileNumber:         e.MobileNumber,
		MobileNumberVerified: e.MobileNumberVerified,
		AltContactNumber:     e.AltContactNumber,
		Email:                e.Email,
		EmailVerified:        e.EmailVerified,
		Locale:               e.Locale,
		IsActive:             e.Active,
		FatherName:           e.FatherName,
		HusbandName:          e.HusbandName,
		Photo:                e.Photo,
		UserID:               e.UserID,
		AdditionalAttributes: e.AdditionalDetails,
		Version:              e.RowVersion,
		RequestID:            e.RequestID,
		AuditDetail:          newAuditDetail(e.CreatedBy, e.ModifiedBy, e.CreatedTime, e.ModifiedTime),
	}
	for i := range e.Addresses {
		if dto := AddressFromEntity(&e.Addresses[i]); dto != nil {
			d.Addresses = append(d.Addresses, *dto)
		}
	}
	for i := range e.Identifiers {
		if dto := IdentifierFromEntity(&e.Identifiers[i]); dto != nil {
			d.Identifiers = append(d.Identifiers, *dto)
		}
	}
	for i := range e.Documents {
		if dto := DocumentFromEntity(&e.Documents[i]); dto != nil {
			d.Documents = append(d.Documents, *dto)
		}
	}
	return d
}

// IndividualsFromEntities converts a slice of entities to a slice of DTOs.
func IndividualsFromEntities(es []Individual) []IndividualDTO {
	out := make([]IndividualDTO, 0, len(es))
	for i := range es {
		if dto := IndividualFromEntity(&es[i]); dto != nil {
			out = append(out, *dto)
		}
	}
	return out
}

// IndividualToEntity converts a wire DTO into a DB entity.
// Server-managed fields (TenantID from header, audit fields from enrichment,
// HashedMobileNumber from encryption) are NOT populated here.
func IndividualToEntity(d *IndividualDTO) *Individual {
	if d == nil {
		return nil
	}
	e := &Individual{
		ID:                   d.ID,
		IndividualID:         d.IndividualID,
		GivenName:            d.GivenName,
		FamilyName:           d.FamilyName,
		OtherNames:           d.OtherNames,
		DateOfBirth:          d.DateOfBirth.ToTimePtr(),
		Gender:               d.Gender,
		Age:                  d.Age,
		MobileNumber:         d.MobileNumber,
		MobileNumberVerified: d.MobileNumberVerified,
		AltContactNumber:     d.AltContactNumber,
		Email:                d.Email,
		EmailVerified:        d.EmailVerified,
		Locale:               d.Locale,
		Active:               d.IsActive,
		FatherName:           d.FatherName,
		HusbandName:          d.HusbandName,
		Photo:                d.Photo,
		UserID:               d.UserID,
		AdditionalDetails:    d.AdditionalAttributes,
		RowVersion:           d.Version,
		RequestID:            d.RequestID,
	}
	for i := range d.Addresses {
		if ent := AddressToEntity(&d.Addresses[i]); ent != nil {
			e.Addresses = append(e.Addresses, *ent)
		}
	}
	for i := range d.Identifiers {
		if ent := IdentifierToEntity(&d.Identifiers[i]); ent != nil {
			e.Identifiers = append(e.Identifiers, *ent)
		}
	}
	for i := range d.Documents {
		if ent := DocumentToEntity(&d.Documents[i]); ent != nil {
			e.Documents = append(e.Documents, *ent)
		}
	}
	return e
}
